fix(alerts&reports): Alerts & Reports will use values from WEBDRIVER_WINDOW option

iercan opened the pull request:
### SUMMARY
This PR resolves #13097
This improvement make thumbnails and new reports module uses WEBDRIVER_WINDOW option from config. 

I defined init function to screenshot object to pass window_size and thumb_size same value for reports so that images will not be shrink . Before It was using harcoded 800x600 value which is defined for thumbnails. 

### TEST PLAN
Create alert and reports. Images should be sent same size defined with WEBDRIVER_WINDOW

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue: 
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

iercan committed:
Merge pull request #1 from apache/master

Sync with master

Commit:
fix: thumbnails and reports will be use WEBDRIVER_WINDOW option

dpgaspar commented:
@iercan thank you for the fix, I restarted CI jobs. But you need to take care of:
https://github.com/apache/superset/pull/13157/checks?check_run_id=1912529461

you can run locally:
```
git add -u
pre-commit
```

iercan commented:
@dpgaspar I executed the commands on my branch. What it is doing?

dpgaspar commented:
@iercan it does a bunch of things, https://github.com/apache/superset/blob/master/.pre-commit-config.yaml
the most relevant are black, mypy, isort

Also take a look at:
https://github.com/apache/superset/blob/master/CONTRIBUTING.md

iercan commented:
@dpgaspar so should I recommit my changes? or running pre-commit is just enough?

Commit:
changes reformatted

iercan commented:
@dpgaspar oh I got it. It reformatted files.

dpgaspar commented on superset/config.py:
```diff
@@ -946,7 +946,7 @@ class CeleryConfig:  # pylint: disable=too-few-public-methods
 
 # Window size - this will impact the rendering of the data
-WEBDRIVER_WINDOW = {"dashboard": (1600, 2000), "slice": (3000, 1200)}
+WEBDRIVER_WINDOW = {"dashboard": (1600, 3000), "slice": (1280, 960)}
```
This is probably a breaking change, why do we need it?

iercan replied:
In my tests too big slice resolution makes pictures blur too much. Here is result for 3000x1200 and 1280x960
![image](https://user-images.githubusercontent.com/3406152/108336911-9703f300-71e5-11eb-8e8d-8f59ec777e74.png)
![image](https://user-images.githubusercontent.com/3406152/108337108-d16d9000-71e5-11eb-962a-9fe2497c62e4.png)

iercan replied:
@dpgaspar For the record. There will be no problem with reports because I provided same thumb_size and window_size for reports. But they still get so blur for thumbnails. I can revert this change if you want.

iercan replied:
I decided to revert that and provide old harcoded sizes for thumbnails so that caches will not be needed to calculated again.

dpgaspar replied:
nice, better to change your PR description: `Additional note: This PR will cause thumbnails recalculate because cache keys depends on window size.
`

iercan replied:
Removed that part

Commit:
config change reverted. thumbnails sizes changed to original

iercan commented on superset/utils/screenshots.py:
```diff
@@ -195,12 +195,34 @@ def resize_image(
+        self.window_size = (
+            window_size or current_app.config["WEBDRIVER_WINDOW"]["slice"]
+        )
+        self.thumb_size = thumb_size or (800, 600)
```
If I didn't understand wrong, screenshots taken with resolution of windows_size and resized to thumb_size(maybe to save space I'm not sure why). I provided thumb_size same with window_size for reports so that they will not shrink.

iercan replied:
It was hardcoded before. I added to init function so that they can be given as different values

vnourdin replied:
OK, thanks :ok_hand:

iercan commented on superset/reports/commands/execute.py:
```diff
@@ -152,11 +152,19 @@ def _get_screenshot(self) -> ScreenshotData:
+            screenshot = ChartScreenshot(
+                url,
+                self._report_schedule.chart.digest,
+                window_size=app.config["WEBDRIVER_WINDOW"]["slice"],
```
Yeah I used config value as default but maybe in future default value could change thats why I thought it is better to provide it from init function.

Commit:
typo fix

Commit:
bugfix

defining defaults in thumbnails.py caused thumbnail caches invalidated.
they moved to init.

dpgaspar reviewed (commented):
@iercan, Although this is a bug, and this PR fixes it, it's also a breaking change. Starting to see the necessity to create 2 new config keys, one for `thumbnails` and another for `reports`, so that it's possible to configure both independently

iercan commented:
> @iercan, Although this is a bug, and this PR fixes it, it's also a breaking change. Starting to see the necessity to create 2 new config keys, one for `thumbnails` and another for `reports`, so that it's possible to configure both independently

Sure I can add 2 different options. But with final edit I made, I don't think it is a breaking change anymore. I keep thumbnails untouched and just reports will use webdriver window. Isn't it WEBDRIVER_WINDOW meant to use just for reports?

dpgaspar commented:
> > @iercan, Although this is a bug, and this PR fixes it, it's also a breaking change. Starting to see the necessity to create 2 new config keys, one for `thumbnails` and another for `reports`, so that it's possible to configure both independently
> 
> Sure I can add 2 different options. But with final edit I made, I don't think it is a breaking change anymore. I kept thumbnails untouched and just reports will use webdriver window. Isn't WEBDRIVER_WINDOW meant to be used just for reports?

it is, it's sort of a breaking change because the new reports will now change its image size. I think it would be more flexible to have the 2 new config keys. Let the new reports still take the screenshot with the previous size, but making it now possible for everyone to configure it

thoughts?

iercan commented:
> it is, it's sort of a breaking change because the new reports will now change its image size. I think it would be more flexible to have the 2 new config keys. Let the new reports still take the screenshot with the previous size, but making it now possible for everyone to configure it
> 
> thoughts?

I think adding 2 different config may get configuration more complicated from point of view of a user. Also I'm not sure thumbnail size should be configurable.

dpgaspar commented:
> > it is, it's sort of a breaking change because the new reports will now change its image size. I think it would be more flexible to have the 2 new config keys. Let the new reports still take the screenshot with the previous size, but making it now possible for everyone to configure it
> > thoughts?
> 
> I think adding 2 different config may get configuration more complicated from point of view of a user. Also I'm not sure thumbnail size should be configurable.

Good point, but still seems weird to me that by configuring `WEBDRIVER_WINDOW` we only end up configuring the new alerts & reports window, while thumbnails is hardcoded. Not explicit nor obvious. But this looks good, and I don't want to stale this fix for long. Thank you for your contribution and patience
